### PR TITLE
Fix bug parsing large streaming JSON objects 🐽

### PR DESCRIPTION
**Pull request by obeattie**
Internally `JsonStreamParser` uses a buffered reader to read the streaming HTTP response. It is possible for a single chunk to exceed the size of this buffer. Previously, if one of these reads resulted in an empty chunk (specifically in the case we saw, the read resulted in just `"\n"`) the stream could get “stuck,” perpetually trying to parse an incomplete object.

In this condition the stream never terminates, causing the k8s namer to stop processing updates to an endpoint until the process is restarted. This was turned up by some internal chaos 🐒 testing, and I think it is quite likely to be related to #1730.

In doing this I have refactored some of the code of `JsonStreamParser`, hopefully making it simpler by removing code that, for example, compensated for [now-fixed bugs in `com.twitter.util`](https://github.com/twitter/util/issues/195).

Tests are included for this and some other previously-uncovered cases. 🔋

### CODE REVIEW

**Comment by wmorgan** [comment]
Amazing. Thanks for this, @obeattie. We'll take a look.

**Comment by obeattie** [comment]
I think the rest failure is unrelated to this change 🤔

**Comment by obeattie** [comment]
I think the test failure is unrelated to this change 🤔

**Comment by deebo91** [comment]
@obeattie I can try rerunning the CI job to see if we run into the same thing.

**Comment by hawkw** [comment]
@obeattie, this is great! I'll get back to you with more detailed review soon. 😄

File: config/src/main/scala/io/buoyant/config/JsonStreamParser.scala [hunk 1]
```diff
@@ -29,105 +28,66 @@ class JsonStreamParser(mapper: ObjectMapper with ScalaObjectMapper) {
...
-      case Throw(e) =>
-        log.warning(e, "json read error")
-    }.handle {
-      case NonFatal(e) => None
-    }
-
-    AsyncStream.fromFuture(read).flatMap {
-      //Fake None element to get around scanLeft being one behind bug
-      //Tracked in https://github.com/twitter/util/issues/195
-      //Can be removed once 195 is fixed
```
Comment (hawkw) [inline config/src/main/scala/io/buoyant/config/JsonStreamParser.scala]: Glad this could be removed! 🙌

File: k8s/src/test/scala/io/buoyant/k8s/JsonTest.scala [hunk 2]
```diff
@@ -17,21 +18,22 @@ class JsonTest extends FunSuite {
 
   test("chunked: message") {
     val obj = TestType("foo", Some(TestType("bar")))
-    val (objs, rest) = Json.readChunked[TestType](Json.writeBuf(obj))
-    assert(objs == Seq(obj))
+    val (decoded, rest) = Json.readBuf[TestType](Json.writeBuf(obj))
+    assert(decoded == Seq(obj))
```
Comment (hawkw) [inline k8s/src/test/scala/io/buoyant/k8s/JsonTest.scala]: I think this renaming makes the tests a bit clearer, thanks! 👍

File: config/src/main/scala/io/buoyant/config/JsonStreamParser.scala [hunk 3]
```diff
@@ -29,105 +28,66 @@ class JsonStreamParser(mapper: ObjectMapper with ScalaObjectMapper) {
...
-          }
-        }
-      )
-      .flatMap(s => AsyncStream.fromSeq(s._1))
-  }
+    // Wrap the reader so reads just terminate if a non-fatal exception occurs. This ensures the below AsyncStream ends
+    // if reads fail
+    val terminatingReader = new Reader {
+      def read(n: Int): Future[Option[Buf]] = reader.read(n).handle {
+        case NonFatal(_) => None
```
Comment (adleong) [inline config/src/main/scala/io/buoyant/config/JsonStreamParser.scala]: What do you think about logging the exception here?

File: config/src/main/scala/io/buoyant/config/JsonStreamParser.scala [hunk 4]
```diff
@@ -29,105 +28,66 @@ class JsonStreamParser(mapper: ObjectMapper with ScalaObjectMapper) {
...
+      val o = parser.getCurrentLocation.getByteOffset.toInt
+      (v, o)
+    } catch {
+      case Incomplete() => (None, 0)
+    } finally {
+      parser.close()
     }
 
-    val rest =
-      if (offset >= chunk.length) Buf.Empty
```
[the remark this answers is not in the document]
Comment (obeattie) [reply config/src/main/scala/io/buoyant/config/JsonStreamParser.scala]: Given the offset is only ever updated to the parser's last location, I don't think this is possible. Even if it were, `Buf.slice`'s doc states that:
>Out of bounds indices are truncated.

**Comment by adleong** [comment]
Hey @obeattie and happy new year!  I'm excited about this fix and I'd like to move forward but I could use some more help understanding the bug that's being fixed here (see my previous comment).

**Comment by obeattie** [comment]
@adleong Sorry, I thought I had already responded to this. If you run the `stream: empty chunk reads` test that I added against the old version of the code, you will see that it fails – specifically, the stream stalls and the test hangs. The new version of the code fixes this.

The underlying reason for this is the old `Incomplete` extractor wouldn't match an exception that was thrown by Jackson: "No content to map due to end-of-input" because it expected such messages to _begin_ "Unexpected end-of-input". This may have worked better in an older version of Jackson but I think it's likely that this always happened, it was just hard to trigger.

We caught this "in the wild" where a HTTP chunk would be terminated with a newline (something that k8s does), with the newline positioned in such a way that it would be read in a separate read operation due to the [buffer size](https://github.com/monzo/linkerd/blob/8f6ae461480fcd5462e87b8fbb88199cbeef9251/k8s/src/main/scala/io/buoyant/k8s/Api.scala#L14). When this case was hit the stream would never recover and no further endpoint updates would be processed.

The rest of the refactor is, in a sense, window dressing, but when I was tracking this down I found the old code a little hard to follow; I think the refactor simplifies things.

**Comment by adleong** [comment]
Ah, makes sense.  I do think your refactors make the code more clear and increase the code quality level.  However, since we currently have active investigations into potential bugs with the k8s namer, I think I'd prefer to make the smallest change possible that fixes the bug.  Even though the code looks solid to me, I'm afraid of changing too many things at a time.
